File: archives/python/1753.py
import sys
input = sys.stdin.readline

import heapq
INF = int(10**9)
V, E = map(int, input().split())
K = int(input())
graph = [[] for _ in range(V+1)]
distance = [INF] * (V+1)

for _ in range(E):
	u, v, w = map(int, input().split())
	graph[u].append((v, w))

def dijkstra(start):
	queue = []
	distance[start] = 0
	heapq.heappush(queue, (0, start))
	while queue:
		dist, node = heapq.heappop(queue)
		if distance[node] < dist:
			continue
		for i in graph[node]:
			cost = dist + i[1]
			if cost < distance[i[0]]:
				distance[i[0]] = cost
				heapq.heappush(queue, (cost, i[0]))

dijkstra(K)
for i in range(1, V+1):
    print("INF" if distance[i] == INF else distance[i])

File: archives/cpp/1753.cpp
#include <bits/stdc++.h>
#define ALL(x) begin(x), end(x);
#define FOR(i, start, end) for (int i = start; i < end; i++)
#define endl "\n"
#define fastio ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
// #pragma GCC optimize("Ofast")
using namespace std;
using ll = long long;
//-------------------------------------
#define INF int(1e9)
using p = pair<int, int>;

int V, E, K;
int dist[20001];
vector<p> graph[20001];
priority_queue<p> pq;

int main() {
    fastio
    
    cin >> V >> E; cin >> K;
    FOR(i, 0, E) {
        int u, v, w; cin >> u >> v >> w;
        graph[u].push_back({v, w});
    }
    fill(dist, dist+20001, INF);

    pq.push({0, K});
    dist[K] = 0;

    while (!pq.empty()) {
        int cost = -pq.top().first;
        int node = pq.top().second;
        pq.pop();

        for (p i: graph[node]) {
            int next_node = i.first;
            int next_cost = i.second;

            if (dist[next_node] > cost + next_cost) {
                dist[next_node] = cost + next_cost;
                pq.push({-dist[next_node], next_node});
            }
        }
    }

    FOR(i, 1, V+1) {
        if (dist[i] == INF) {
            cout << "INF" << endl;
        } else {
            cout << dist[i] << endl;
        }
    }

    return 0;
} 